### ISSUE TITLE
Split documentations and implementation

### ISSUE THREAD

**Issue by m-mohr** [Open-EO/openeo-udf#10]
I think it would be a good idea to separate this repository into three parts:
1. A user guide how to use and code UDFs (e.g. an extended version of the currently available "Coding UDFs" for R and Python, could be part of the final user guide we also need to separate from the API specification).
2. A implementors guide (i.e. the documentation we have now, potentially in this repository or the openeo-api as extension)
3. The Python reference implementation + documentation specific to this implementation (similar to how the r-udf is separate as well, so for example into a openeo-python-udf repository.)

Issue originated from: https://github.com/Open-EO/openeo-processes/issues/42

**Comment by huhabla** [comment]
I think splitting the documentation makes it harder vor potential developers to understand the whole concept. It must be clear howto deploy a UDF server or howto use the command line tool to implement a UDF. There is a detailed API description available based on the source code.

**Comment by m-mohr** [comment]
So users are expected to read the whole developer docs, too?
Where can I point a user that wants to simply write a UDF?

**Comment by huhabla** [comment]
Yes, this is required. How should the user otherwise test his UDF while developing it? IMHO is the openEO API that supports UDF via graphs is not well designed for UDF development. Howto debug your UDF code when you are using a REST API?

**Comment by huhabla** [comment]
The easiest solution for UDF coding ist to use this repo fro development and testing. New UDF function can be added in a local branch and unit tests can be implemented for them. Since UDF's should work on any backend, the development should be independent of it.

**Comment by jdries** [comment]
I had some remarks/questions in the same direction at the EODC forum today.
One approach I was thinking of is integrating of udf debugging into the client.
Basically the client would have to download a piece of sample data, and then pass it into the UDF in exactly the same way as the backends would do it.

**Comment by huhabla** [comment]
Users who just want to use an UDF in a process graph, should not care about any UDF implementation or setup concepts. What these users need is a database of available UDF functions with detailed description and the documentation howto put UDF's into the process graph. The best approach would be if the user can specify a URL that points to a UDF in the client or a file that will be uploaded by the openEO client to the backend. Maybe the backend can provide backend data specific UDF's?

**Comment by flahn** [comment]
I think the issue @m-mohr raised in the beginning aimed more towards the general UDF API being a stand-alone repository (like openEO API) that contains the OpenAPI documentation in order to help back-end developers to find information how to interact with UDF services.
The UDF implementations for python, R and potentially others should then contain examples and user guidelines how to develop scripts, that can be shared later via the openEO hub.

@jdries I think, having sample data for UDF script development is a good approach to tackle this problem, since you can only really test the UDF code with correct data and we should push this forward. Maybe you can raise another issue for this, because we drift away from the topic of this issue.

@huhabla you are right, in the end it should be possible for openeo users to use publicly available UDF scripts in a similar fashion as processes offered by the back-ends and this might work via pointing to the URL of the script on the openEO hub.
